### PR TITLE
8295713: runtime/ParallelLoad/SuperWait/SuperWaitTest.java fails intermittently on Windows

### PR DESCRIPTION
**Pull request by coleenp**
I added another semaphore so the test will run in the order I want it to.  Alternate suggestions or corrections welcome.
Running with GHA, and our tier1 in progress.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295713](https://bugs.openjdk.org/browse/JDK-8295713): runtime/ParallelLoad/SuperWait/SuperWaitTest.java fails intermittently on Windows

### Reviewers
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) ⚠️ Review applies to [6c4eced2](https://git.openjdk.org/jdk/pull/10822/files/6c4eced275a5705710eb7c5d9a05db8ca0d0cd5d)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [8278537a](https://git.openjdk.org/jdk/pull/10822/files/8278537a1c9d6202ff0669563693b5eba99b063e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10822/head:pull/10822` \
`$ git checkout pull/10822`

Update a local copy of the PR: \
`$ git checkout pull/10822` \
`$ git pull https://git.openjdk.org/jdk pull/10822/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10822`

View PR using the GUI difftool: \
`$ git pr show -t 10822`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10822.diff">https://git.openjdk.org/jdk/pull/10822.diff</a>

</details>

### CODE REVIEW

**Review by pchilano — APPROVED** [review]
Looks good to me. But I think mainSync is not needed now. Thread 1 will always get to wait() before Thread 2 reaches the notify() call while loading D, so that should be enough to avoid the lost notification issue.

**Comment by coleenp** [comment]
Thanks Patricio, I don't need mainSync anymore because the first thread will wait.

**Review by dholmes-ora — COMMENTED** [review]
Your are doing an unconditional `wait()` which in theory could be affected by a spurious wakeup. If you actually do:
```
while (!DisLoading) {
  wait();
}
```
with a corresponding
```
DIsLoading = true;
notify();
```
then I think you could dispense with the semaphore altogether.

**Comment by coleenp** [comment]
I need to both handle the spurious wakeup and the case where the second thread loads everything, and then the first thread then waits, after D's notification - the missed notification.

**Comment by dholmes-ora** [comment]
> I need to both handle 1. the spurious wakeup from wait(), and 2. the case where the second thread loads everything, and then the first thread then waits, after D's notification - the missed notification.

That is what `DIsLoaded` indicates. There's no missed notification because you won't `wait` if D is already loaded.

**Comment by coleenp** [comment]
It won't test what I'm trying to test though, but maybe it's ok.  I added the test for illustrative purposes.

File: test/hotspot/jtreg/runtime/ParallelLoad/SuperWait/SuperWaitTest.java [hunk 1]
```diff
@@ -36,19 +36,17 @@
 
 public class SuperWaitTest {
 
-    private static Semaphore mainSync = null;
+    private static volatile boolean dIsLoading = false;
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ParallelLoad/SuperWait/SuperWaitTest.java:39]: This doesn't need to be volatile as it is only accessed under the classloader lock.

**Review by dholmes-ora — COMMENTED** [review]
Sorry Coleen some of the intermediate changes here confused me. The original use of `mainSync` is slowing down thread1 (loading A) so that it can't proceed until thread 2 starts to load C and has the lock of L2 (which sets up the potential deadlock scenario). That could be added back in conjunction with the `dIsLoaded` usage. In both cases thread 2 could always get to go first and load C and D before thread 1 has a chance to do anything.

**Comment by coleenp** [comment]
My n-1 change makes thread 2 wait for loading C and D until thread1 is in the locked region for loading A (which was the deadlock scenario before the wait() is introduced).   It doesn't really matter if thread 2 goes first with the dIsLoading change.  Either way, the test will complete and not miss a notify.

File: test/hotspot/jtreg/runtime/ParallelLoad/SuperWait/SuperWaitTest.java (hunk 1 above)
Comment (coleenp) [reply test/hotspot/jtreg/runtime/ParallelLoad/SuperWait/SuperWaitTest.java:39]: ThreadPrint.println("Waiting for " + name);
                    while (!dIsLoading) {  // guard against spurious wakeup
                        try {
                            wait(); // let the other thread have this lock.
                        } catch (InterruptedException ie) {}
                    }

Do we not have to make dIsLoading volatile?

File: test/hotspot/jtreg/runtime/ParallelLoad/SuperWait/SuperWaitTest.java [hunk 2]
```diff
@@ -36,19 +36,17 @@
 
 public class SuperWaitTest {
 
-    private static Semaphore mainSync = null;
+    private static boolean dIsLoading = false;
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ParallelLoad/SuperWait/SuperWaitTest.java:39]: Nit: sorry just noticed the placement here. This is really an instance field of MyLoaderOne.

**Review by dholmes-ora — APPROVED** [review]
Okay this seems fine from a logic perspective. There's one nit below that I missed before but up to you whether to change it.
Thanks.

File: test/hotspot/jtreg/runtime/ParallelLoad/SuperWait/SuperWaitTest.java (hunk 2 above)
Comment (coleenp) [reply test/hotspot/jtreg/runtime/ParallelLoad/SuperWait/SuperWaitTest.java]: I fixed it.

**Comment by coleenp** [comment]
Thanks for the reviews Patricio and David.
/integrate

**Pull request closed without merging**